server/webhook_notifier.go

```diff
@@ -0,0 +1,157 @@
+package server
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"log"
+	"net"
+	"net/http"
+	"time"
+)
+
+// WebhookNotifier implements ConnectionNotifier by sending a POST request to a webhook URL.
+// The payload is a JSON object with the following fields:
+//
+//	event: the event type, one of "connecting", "missing-backend", "failed-backend-connection", "success"
+//	status: the status of the event, one of "missing-backend", "failed-backend-connection", "success"
+//	client-host: the client host
+//	client-port: the client port
+//	server-address: the server address
+//	player-info: the player info, if any
+//	backend: the backend host and port, if any
+//	error: the error, if any
+type WebhookNotifier struct {
+	url         string
+	requireUser bool
+
+	client *http.Client
+}
+
+const (
+	WebhookEventConnecting = "connecting"
+)
+
+const (
+	WebhookStatusMissingBackend          = "missing-backend"
+	WebhookStatusFailedBackendConnection = "failed-backend-connection"
+	WebhookStatusSuccess                 = "success"
+)
+
+type WebhookNotifierPayload struct {
+	Event           string      `json:"event"`
+	Status          string      `json:"status"`
+	ClientHost      string      `json:"client-host"`
+	ClientPort      int         `json:"client-port"`
+	ServerAddress   string      `json:"server-address"`
+	PlayerInfo      *PlayerInfo `json:"player-info,omitempty"`
+	BackendHostPort string      `json:"backend,omitempty"`
+	Error           string      `json:"error,omitempty"`
+}
+
+func NewWebhookNotifier(url string, requireUser bool) *WebhookNotifier {
+
+	return &WebhookNotifier{
+		url:         url,
+		requireUser: requireUser,
+		client: &http.Client{
+			Timeout: 30 * time.Second,
+		},
+	}
+}
+
+func (w *WebhookNotifier) NotifyMissingBackend(ctx context.Context, clientAddr net.Addr, serverAddress string, playerInfo *PlayerInfo) error {
+	if w.requireUser && playerInfo == nil {
+		return nil
+	}
+
+	payload := &WebhookNotifierPayload{
+		Event:         WebhookEventConnecting,
+		Status:        WebhookStatusMissingBackend,
+		ClientHost:    clientAddr.(*net.TCPAddr).IP.String(),
+		ClientPort:    clientAddr.(*net.TCPAddr).Port,
+		ServerAddress: serverAddress,
+		PlayerInfo:    playerInfo,
+		Error:         "No backend found",
+	}
+
+	return w.send(ctx, payload)
+}
+
+func (w *WebhookNotifier) NotifyFailedBackendConnection(ctx context.Context, clientAddr net.Addr, serverAddress string,
```

Comment:
There has to be a better way to do this. You shouldn't have functions that return different responses. The best case should have one function and when its called it should provide only the information it has at the time.

You're kind of repeating yourself here.

Reply:
That's a bit rude.  I thought API was a balance of cleanly documented yet filling out respective parts of the payload.

Reply:
> That's a bit rude. 

I apologize, I did not mean to have it sound like that. Working with Kubernetes + YAML, the goal was always to avoid repeating yourself wherever possible. 

The best approach would be to create a object that is effectively "empty" and fill in information as you have it. When it's sent to the audit, if it's not filled out the server did not have it.

Reply:
Yeah, I've been developing code for 30 years...I know about DRY 😄I'll probably collapse the two failure modes into one method, but my coding preference won't allow much more than that.

Reply:
Since I have designed the `ConnectionNotifier` interface to be generic, combining those methods would instead require the caller to pass in a "status string". Maybe there's a Discord notifier and it wants to use an emoji instead of a particular string to convent a missing backend.

When you've written code as long as I have you start to find the balances between repeating a few lines of code and the maintainability and extensibility of the code around it.